lib/Index/IndexDatastore.cpp

```diff
@@ -707,28 +712,29 @@ bool IndexDatastoreImpl::init(IndexStoreRef idxStore,
   auto UnitRepo = std::make_shared<StoreUnitRepo>(this->IdxStore, SymIndex, Delegate, CanonPathCache);
   std::weak_ptr<StoreUnitRepo> WeakUnitRepo = UnitRepo;
   auto OnUnitsChange = [WeakUnitRepo, Delegate](IndexStore::UnitEventNotification EventNote) {
-    auto UnitRepo = WeakUnitRepo.lock();
-    if (!UnitRepo)
-      return;
-
-    std::vector<UnitEventInfo> evts;
+    // Pass each registration event as a separate block to be processed by the global serial queue.
+    // This allows intermixing processing of registration events from multiple workspaces.
     for (size_t i = 0, e = EventNote.getEventsCount(); i != e; ++i) {
-      auto evt = EventNote.getEvent(i);
-      evts.push_back(UnitEventInfo{evt.getKind(), evt.getUnitName()});
-    }
-
-    // Create the block with QoS explicitly to ensure that the QoS from the indexstore callback can't affect the onFilesChange priority. This call may do a lot of I/O and we don't want to wedge the system by running at elevated priority.
-    dispatch_block_t onUnitChangeBlock = dispatch_block_create_with_qos_class(DISPATCH_BLOCK_INHERIT_QOS_CLASS, unitChangesQOS, 0, ^{
-      Delegate->processingAddedPending(evts.size());
-      UnitRepo->onFilesChange(std::move(evts), [&](unsigned NumCompleted){
-        Delegate->processingCompleted(NumCompleted);
-      }, [&](){
-        // FIXME: the database should recover.
+      auto cevt = EventNote.getEvent(i);
+      UnitEventInfo evt = UnitEventInfo{cevt.getKind(), cevt.getUnitName()};
+
+      // Create the block with QoS explicitly to ensure that the QoS from the indexstore callback can't affect the onFilesChange priority. This call may do a lot of I/O and we don't want to wedge the system by running at elevated priority.
+      dispatch_block_t onUnitChangeBlock = dispatch_block_create_with_qos_class(DISPATCH_BLOCK_INHERIT_QOS_CLASS, unitChangesQOS, 0, ^{
+        auto UnitRepo = WeakUnitRepo.lock();
+        if (!UnitRepo)
+          return;
+
+        Delegate->processingAddedPending(1);
```

Comment:
Putting each event in its own block is unlikely to intermix the events between workspaces, because it still queues up all of the blocks at the beginning using `dispatch_async`.  Unless the workspaces are racing to add events, all the events from workspace 1 get queued before workspace 2 even starts.

If we want to intermix the events, we probably need to be more proactive.  Each workspace could have a list of outstanding events, and the onUnitChange block can pull 1 (or any batch size) events from there at a time and reschedule itself with dispatch_async.  That way other workspaces can add their own block to the queue in between handing events.

Reply:
Good point, see changes.